ambari-server/src/test/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessorTest.java

```diff
@@ -2168,6 +2177,198 @@ public void testDoUpdateForClusterCreate_SingleHostProperty__defaultValue() thro
+    Map<String, String> myserviceSite = hostGroupConfig.getProperties().get("myservice-site");
+    assertEquals(myserviceSite, ImmutableMap.of("myservice_slave_address", "testhost:8080"));
+  }
+
+  @Test
+  public void testHostgroupUpdater() throws Exception {
+    Set<String> components = ImmutableSet.of("NAMENODE", "SECONDARY_NAMENODE", "RESOURCEMANAGER");
+    Configuration hostGroupConfig = new Configuration(emptyMap(), emptyMap());
+    TestHostGroup group1 = new TestHostGroup("master1", components, ImmutableList.of("master1_host"), hostGroupConfig);
+    TestHostGroup group2 = new TestHostGroup("master2", components, ImmutableList.of("master2_host"), hostGroupConfig);
+    TestHostGroup group3 =
+      new TestHostGroup("master3", components, ImmutableList.of("master3_host1", "master3_host2"), hostGroupConfig);
+    Configuration clusterConfig = new Configuration(new HashMap<>(), new HashMap<>());
+    ClusterTopology topology = createClusterTopology(bp, clusterConfig, ImmutableList.of(group1, group2, group3));
+
+    BlueprintConfigurationProcessor.HostGroupUpdater updater = BlueprintConfigurationProcessor.HostGroupUpdater.INSTANCE;
+
+    assertEquals(
+      "master1_host:8080",
+      updater.updateForClusterCreate(
+        "mycomponent.url",
+        "%HOSTGROUP::master1%:8080",
+        clusterConfig.getProperties(),
+        topology));
+
+    assertEquals(
+      "master1_host:8080,master2_host:8080",
+      updater.updateForClusterCreate(
+        "mycomponent.urls",
+        "%HOSTGROUP::master1%:8080,%HOSTGROUP::master2%:8080",
+        clusterConfig.getProperties(),
+        topology));
+
+    assertThat(
+      updater.updateForClusterCreate(
+        "mycomponent.urls",
+        "%HOSTGROUP::master3%:8080,%HOSTGROUP::master3%:8080",
+        clusterConfig.getProperties(),
+        topology),
+      anyOf(is("master3_host1:8080,master3_host2:8080"), is("master3_host2:8080,master3_host1:8080")));
+
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testHostgroupUpdater_NonExistingHostgroup() throws Exception {
+    Set<String> components = ImmutableSet.of("NAMENODE", "SECONDARY_NAMENODE", "RESOURCEMANAGER");
+    Configuration hostGroupConfig = new Configuration(emptyMap(), emptyMap());
+    TestHostGroup group1 = new TestHostGroup("master1", components, ImmutableList.of("master1_host"), hostGroupConfig);
+    Configuration clusterConfig = new Configuration(new HashMap<>(), new HashMap<>());
+    ClusterTopology topology = createClusterTopology(bp, clusterConfig, ImmutableList.of(group1));
+    BlueprintConfigurationProcessor.HostGroupUpdater updater = BlueprintConfigurationProcessor.HostGroupUpdater.INSTANCE;
+
+    updater.updateForClusterCreate(
+      "mycomponent.urls",
+      "%HOSTGROUP::master2%:8080",
+      clusterConfig.getProperties(),
+      topology);
+  }
+
+  @Test(expected = IllegalStateException.class)
```

Comment:
Minor issue:  Where is this check occurring, and where is this exception thrown?  I've looked through this patch a few times, but I wasn't able to determine where the check for multiple instances of a given hostgroup name occurring in a property value.  

I'm probably just missing something, so if you could clarify this it would be a big help.